### PR TITLE
Feature/patch bom alternative

### PR DESCRIPTION
**Pull request by elomagic**

### CODE REVIEW

File: core/src/main/java/de/medavis/lct/core/patcher/ComponentMetaDataManager.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        clear();
+        this.componentMetaDataList = List.copyOf(list);
+    }
+
+    public void load(@NotNull URI uri) {
+        LOGGER.info("Loading custom rules from '{}'.", uri);
+
+        try {
+            ComponentMetaDataLoader loader = new ComponentMetaDataLoader();
+            if ("file".equals(uri.getScheme())) {
```
Comment (martinleim) [inline core/src/main/java/de/medavis/lct/core/patcher/ComponentMetaDataManager.java:63]: Both branches are identical

File: core/src/main/java/de/medavis/lct/core/patcher/ComponentMetaDataManager.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+     * @return Returns true, when unsupported warnings was found
+     */
+    public boolean validateLicenseMappedNames(@NotNull Set<String> supportedLicenseNames) {
+        AtomicBoolean result = new AtomicBoolean(false);
+
+        componentMetaDataList
+                .stream()
+                .filter(cm -> StringUtils.isNotBlank(cm.mappedName()) && !supportedLicenseNames.contains(cm.mappedName()))
+                .forEach(cm -> {
+                    result.set(true);
```
Comment (martinleim) [inline core/src/main/java/de/medavis/lct/core/patcher/ComponentMetaDataManager.java:97]: This is a weird use of an `AtomicBoolean`. Also, it would be more convenient to have a single log line listing all warnings. Suggestion:

```java
var componentsWithInvalidLicenses = componentMetaDataList.stream()
                                           .filter(cm -> StringUtils.isNotBlank(cm.mappedName()) && !supportedLicenseNames.contains(cm.mappedName()))
                                           .map(ComponentMetadata::mappedName)
                                           .collect(Collectors.toList());
if (!componentsWithInvalidLicenses.isEmpty()) {
    LOGGER.warn("Your component meta data configuration contains unsupported SPDX names: '{}'.", componentsWithInvalidLicenses);
}
return componentsWithInvalidLicenses.isEmpty();
```

File: core/src/main/java/de/medavis/lct/core/patcher/BomPatcher.java [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+                        url -> spdxLicenseManager.load(URI.create(url.toString())),
+                        spdxLicenseManager::loadDefaults
+                );
+
+        configuration
+                .getComponentMetadataUrl()
+                .map(url -> URI.create(url.toString()))
+                .ifPresent(componentMetaDataManager::load);
+
+        componentMetaDataManager.validateLicenseMappedNames(spdxLicenseManager.getSupportedLicenseNames());
```
Comment (martinleim) [inline core/src/main/java/de/medavis/lct/core/patcher/BomPatcher.java:88]: What's the point in validating the licenses if you don't care about the result?

File: core/src/main/java/de/medavis/lct/core/patcher/BomPatcher.java [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+            }
+
+            String originalBom = BomGeneratorFactory.createJson(version, bom).toJsonString();
+
+            Optional.ofNullable(bom.getComponents())
+                    .orElse(List.of())
+                    .forEach(component -> {
+                        String purl = Objects.toString(component.getPurl(), "");
+
+                        if (component.getLicenses() == null) {
```
Comment (martinleim) [inline core/src/main/java/de/medavis/lct/core/patcher/BomPatcher.java:145]: The code is pretty hard to understand. I am not sure if I understand that right:
- If there are no licenses in the BOM, write the ones from component meta data into the BOM
-  If there are licenses in the BOM, check if they conform with the SPDX list, but otherwise ignore them and do the same as in the case above.

Is that correct? It does not really make sense to me... Additionally, it only takes into account the component meta data, but not the license mappings.
What would make sense to me would be the following algorithm:
- Transform component from BOM into `ComponentMetaData, enriching or replacing the licenses from those configured in component metadata, license mappings and license list. This logic is already implemented in `de.medavis.lct.core.list.ComponentLister#enrichWithMetadata` and could be reused.
- Write the licenses from ComponentMetadata into the BOM component

**Review by martinleim — COMMENTED** [review]
I am not fully through, but I think we should discuss the overall approach of the patch process.

File: core/src/main/java/de/medavis/lct/core/Json5MapperFactory.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import com.fasterxml.jackson.core.json.JsonReadFeature;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Factory for creating JSON5 Mappers
+ */
+public class Json5MapperFactory {
```
[the remark this answers is not in the document]
Comment (elomagic) [reply core/src/main/java/de/medavis/lct/core/Json5MapperFactory.java:32]: A convenient feature like in properties or XML files to comment and understand settings

File: core/src/main/java/de/medavis/lct/core/metadata/ComponentMetaDataLoader.java [hunk 6]
```diff
@@ -20,24 +20,31 @@
...
 import org.slf4j.LoggerFactory;
 
 public class ComponentMetaDataLoader {
 
-    private static final Logger logger = LoggerFactory.getLogger(ComponentMetaDataLoader.class);
+    private static final Logger LOGGER = LoggerFactory.getLogger(ComponentMetaDataLoader.class);
 
-    public Collection<ComponentMetadata> load(URL metadataUrl) {
+    public Collection<ComponentMetadata> load(@NotNull URL metadataUrl) {
         ObjectMapper objectMapper = new ObjectMapper();
```
[the remark this answers is not in the document]
Comment (elomagic) [reply core/src/main/java/de/medavis/lct/core/metadata/ComponentMetaDataLoader.java]: Fixed

File: core/src/main/java/de/medavis/lct/core/patcher/AbstractRestClient.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+import java.io.IOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.List;
+
+public abstract class AbstractRestClient {
```
[the remark this answers is not in the document]
Comment (elomagic) [reply core/src/main/java/de/medavis/lct/core/patcher/AbstractRestClient.java:39]: Fixed

File: core/src/main/java/de/medavis/lct/core/patcher/BomPatcher.java [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+    /**
+     * Patches a BOM file.
+     *
+     * @param sourceFile Source BOM file
+     * @param targetFile Target BOM file
+     * @return Returns true if patching was successful. False, when BOM leaved untouched.
+     */
+    public boolean patch(@NotNull Path sourceFile, @NotNull Path targetFile) {
+        try (InputStream in = Files.newInputStream(sourceFile)) {
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
```
[the remark this answers is not in the document]
Comment (elomagic) [reply core/src/main/java/de/medavis/lct/core/patcher/BomPatcher.java]: Fixed

File: core/src/main/java/de/medavis/lct/core/patcher/ComponentMetaDataManager.java (hunk 1 above)
Comment (elomagic) [reply core/src/main/java/de/medavis/lct/core/patcher/ComponentMetaDataManager.java]: Fixed

File: core/src/main/java/de/medavis/lct/core/patcher/ComponentMetaDataManager.java (hunk 2 above)
Comment (elomagic) [reply core/src/main/java/de/medavis/lct/core/patcher/ComponentMetaDataManager.java]: Fixed

File: core/src/main/java/de/medavis/lct/core/patcher/BomPatcher.java (hunk 3 above)
Comment (elomagic) [reply core/src/main/java/de/medavis/lct/core/patcher/BomPatcher.java]: Renamed and fixed

**Review by martinleim — CHANGES_REQUESTED** [review]
Hi @elomagic, summary of our discussion:
Patcher will be reworked so that it
1. Read components from BOM
1. Compares each component against the configured meta data. If there is a match, check if has licenses configured and replace the licenses from BOM wth those from the configured meta data
2. Normalize license name against those configured in license mapping data
3. Use license IDs and URL from configured license data
4. Patch BOM by overwriting the licenses of each component with the licenses obtained from the previous steps.

File: core/src/main/java/de/medavis/lct/core/patcher/AbstractRestClient.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
```
Comment (martinleim) [inline core/src/main/java/de/medavis/lct/core/patcher/AbstractRestClient.java:39]: This class is a generic utility that is used beyond the scope of BOM patching. It should be in a top-level package on its own.

File: core/src/main/java/de/medavis/lct/core/patcher/SpdxLicenseManager.java
Comment (martinleim) [inline core/src/main/java/de/medavis/lct/core/patcher/SpdxLicenseManager.java:1]: This class is part of the BOM patcher package. However, its configuration is part of the global `Configuration` option. This is a mismatch.

Since I can imagine that we want to use these SDPX license IDs in other contexts than BOM patching as well, I would propose to move this class (in modified form, see comment in `BomPatcher`) together with the classes from the `model` subpackage into a top-level package on its own.

File: core/src/main/resources/de/medavis/lct/core/patcher/DefaultLicenseMapping.json5
Comment (martinleim) [inline core/src/main/resources/de/medavis/lct/core/patcher/DefaultLicenseMapping.json5:1]: File is not used and can be deleted.

File: core/src/main/resources/de/medavis/lct/core/patcher/SpdxLicenseList.json5
Comment (martinleim) [inline core/src/main/resources/de/medavis/lct/core/patcher/SpdxLicenseList.json5:1]: Since this file is fetched during the build process, we can remove it as well.

File: core/src/test/java/de/medavis/lct/core/patcher/ComponentMetaDataManagerTest.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+import static org.junit.jupiter.api.Assertions.*;
+
+@ExtendWith(MockitoExtension.class)
+class ComponentMetaDataManagerTest {
+
+    @Mock()
+    private Logger mockedLogger;
+
+    @Test
+    void testFindMatch() throws URISyntaxException {
```
Comment (martinleim) [inline core/src/test/java/de/medavis/lct/core/patcher/ComponentMetaDataManagerTest.java:43]: This test tests code that is not used in production. Suggestion: Remove the tested method and the test.

File: core/src/test/resources/de/medavis/lct/core/patcher/test-rules.json5
Comment (martinleim) [inline core/src/test/resources/de/medavis/lct/core/patcher/test-rules.json5:1]: This file is not used and can be removed.

File: jenkins/src/main/java/de/medavis/lct/jenkins/patch/BomPatcherBuilder.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        return outputFile;
+    }
+
+    @Override
+    public void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNull EnvVars env, @NonNull Launcher launcher, @NonNull TaskListener listener)
+            throws AbortException, InterruptedException {
+        try {
+            final JenkinsLogger logger = new JenkinsLogger(listener);
+            logger.info("Patching BOM from %s into %s.%n", inputFile, outputFile);
+            // TODO Logger?
```
Comment (martinleim) [inline jenkins/src/main/java/de/medavis/lct/jenkins/patch/BomPatcherBuilder.java:56]: Remove TOOD and commented code

File: core/src/main/java/de/medavis/lct/core/Json5MapperFactory.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (martinleim) [reply core/src/main/java/de/medavis/lct/core/Json5MapperFactory.java:32]: Nice, I like it,
